### PR TITLE
fix(pnpm): defer lock file constraint parsing

### LINKED ISSUE #22250: PNPN engine constraint
### How are you running Renovate?

Mend Renovate hosted app on github.com

### If you're self-hosting Renovate, tell us what version of Renovate you run.

_No response_

### If you're self-hosting Renovate, select which platform you are using.

None

### Was this something which used to work for you, and then stopped?

It used to work, and then stopped

### Describe the bug

There is some issue with the way Renovate uses the PNPM engine constraint extracted from `package.json`. Or maybe even with the actual extraction of the constraint we are not sure yet. 

Below are two user submitted reports of the bugs:

1) Renovate uses the latest pnpm version instead of the one from `engines` in `package.json`, when updating a dependency. [discussion](https://github.com/renovatebot/renovate/discussions/22140)
_Note:  In the same repo, Renovate correctly installed latest version of pnpm when updating `pnpm`._ 

> Reproduction: https://github.com/Exlint/dashboard/pull/1055

2) Renovate used extracted pnpm version from `engines` in `package.json`, but it did so when updating the `pnpm` version which is not desired. [disucssion](https://github.com/renovatebot/renovate/discussions/22226)
_Note:   In the same repo, Renovate correctly installed latest version of pnpm when updating other dependencies ._ 

>Reproduction: https://github.com/nisheretail/renovate-pnpm-artifact-issue/pull/5

Seems like Renovate gets stuck with using one pnpm version for all the updates in a repo. ie. In case-1 regardless of the type of update pnpm's latest version is being installed whereas for case-2 the version from `engines` is being used.

### Relevant debug logs

<details><summary>Logs</summary>

```
Copy/paste the relevant log(s) here, between the starting and ending backticks
```

</details>

### Have you created a minimal reproduction repository?

I have linked to a minimal reproduction in the description above

### PR DESCRIPTION
**Pull request by rarkins**
## Changes

Moves lock file constraint parsing from extract phase to post-update.

## Context

- Fixes #22250

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/modules/manager/npm/post-update/pnpm.ts [hunk 1]
```diff
@@ -153,3 +153,63 @@ async function getPnpmConstraint(
   }
   return result;
 }
+
+export function getConstraintsFromLockFile(lockfileVersion: number): string {
```
Comment (viceice) [inline lib/modules/manager/npm/post-update/pnpm.ts:157]: where does it get called? 🤔

shouldn't it be called in function above where also lockfile is handled?

**Comment by RahulGautamSingh** [comment]
You moved constraint extraction to post-update so that when we extract pnpm version from `package.json` we get the updated version incases where pnpm is getting updated. But how does it help with the cases where pnpm isn't updated shouldn't it already be working correctly for those cases. ie this case https://github.com/renovatebot/renovate/discussions/22140 

TLDR; Did you figure out why it wasn't working correctly for https://github.com/renovatebot/renovate/discussions/22140

**Comment by rarkins** [comment]
I worked out why it wasn't working properly - it's due to monorepos. The "child" package.json have no engines defined. During extraction the get given constraints.pnpm due to the lock file, and then that essentially overrides the more exact pnpm constraint from the root package.json engines

File: lib/modules/manager/npm/range.ts [hunk 2]
```diff
@@ -19,7 +19,7 @@ export function getRangeStrategy(config: RangeConfig): RangeStrategy {
   }
   if (depType === 'peerDependencies') {
     // Widen peer dependencies
-    logger.debug('Widening peer dependencies');
+    logger.trace({ currentValue }, 'Widening peer dependencies');
```
Comment (rarkins) [inline lib/modules/manager/npm/range.ts:22]:
```suggestion
    logger.debug('Widening peer dependencies');
```

File: lib/modules/manager/npm/post-update/pnpm.ts [hunk 3]
```diff
@@ -41,9 +41,10 @@ export async function generateLockFile(
     const pnpmToolConstraint: ToolConstraint = {
       toolName: 'pnpm',
       constraint:
-        getPnpmConstraintFromUpgrades(upgrades) ??
-        config.constraints?.pnpm ??
-        (await getPnpmConstraint(lockFileDir)),
+        getPnpmConstraintFromUpgrades(upgrades) ?? // if pnpm is being upgraded, it comes first
+        config.constraints?.pnpm ?? // from user config or extraction
```
Comment (viceice) [inline lib/modules/manager/npm/post-update/pnpm.ts:45]: on other managers we always prefer user config first. 🤔

**Review by viceice — COMMENTED** [review]
otherwise LGTM

File: lib/modules/manager/npm/post-update/pnpm.ts (hunk 3 above)
Comment (rarkins) [reply lib/modules/manager/npm/post-update/pnpm.ts:45]: I think that's a mistake if we're explicitly upgrading `pnpm` in the PR. Also it won't work for pnpm because `constraints` includes extracted constraints from the `package.json` which will then be outdated. It would potentially be ok if `constraints` were only user-configured constraints but in this case it's not

**Review by RahulGautamSingh — DISMISSED** [review]

File: lib/modules/manager/npm/post-update/pnpm.ts (hunk 3 above)
Comment (viceice) [reply lib/modules/manager/npm/post-update/pnpm.ts:45]: ok, so this works on pnpm update, but will fail all following updates when merged to main, because the user constraints are probably not yet updated.

**Review by viceice — DISMISSED** [review]
we can improve in a follow up PR

File: lib/modules/manager/npm/post-update/pnpm.ts [hunk 4]
```diff
@@ -129,27 +130,89 @@ async function getPnpmConstraint(
...
   }
-  return result;
+  return constraint;
 }
+
+/**
+ pnpm lockfiles have corresponding version numbers called "lockfileVersion"
+ each lockfileVersion can only be generated by a certain pnpm version ranges
+ eg. lockfileVersion: 5.4 can only be generated by pnpm version >=7 && <8
+ official list can be found here : https:github.com/pnpm/spec/tree/master/lockfile
```
Comment (viceice) [inline lib/modules/manager/npm/post-update/pnpm.ts:182]:
```suggestion
 official list can be found here : https://github.com/pnpm/spec/tree/master/lockfile
```

File: lib/modules/manager/npm/post-update/pnpm.ts [hunk 5]
```diff
@@ -129,27 +130,89 @@ async function getPnpmConstraint(
...
-      if (
-        is.number(pnpmLock.lockfileVersion) &&
-        pnpmLock.lockfileVersion < 5.4
-      ) {
-        result = '<7';
-      }
+  return constraint;
+}
+
+export async function getConstraintFromLockFile(
```
Comment (viceice) [inline lib/modules/manager/npm/post-update/pnpm.ts:145]: export for tests?

Comment (rarkins) [reply lib/modules/manager/npm/post-update/pnpm.ts:145]: Yes

**Review by viceice — APPROVED** [review]
